airbyte-webapp/src/components/connection/CatalogTree/next/useCatalogTreeTableRowProps.tsx

```diff
@@ -0,0 +1,107 @@
+/* eslint-disable css-modules/no-unused-class */
+import { faMinus, faPlus } from "@fortawesome/free-solid-svg-icons";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import classNames from "classnames";
+import { useField } from "formik";
+import { isEqual } from "lodash";
+import { useMemo } from "react";
+
+import { ModificationIcon } from "components/icons/ModificationIcon";
+import { PillButtonVariant } from "components/ui/PillSelect/PillButton";
+
+import { SyncSchemaStream } from "core/domain/catalog";
+import { useBulkEditSelect } from "hooks/services/BulkEdit/BulkEditService";
+import { useConnectionFormService } from "hooks/services/ConnectionForm/ConnectionFormService";
+
+import styles from "./CatalogTreeTableRow.module.scss";
+
+export const useCatalogTreeTableRowProps = (stream: SyncSchemaStream) => {
+  const { initialValues } = useConnectionFormService();
+  const [isSelected] = useBulkEditSelect(stream.id);
+
+  const [, { error }] = useField(`schema.streams[${stream.id}].config`);
+
+  // in case error is an empty string
+  const hasError = error !== undefined;
+
+  const isStreamEnabled = stream.config?.selected;
+
+  const statusToDisplay = useMemo(() => {
+    const rowStatusChanged =
+      initialValues.syncCatalog.streams.find(
+        (item) => item.stream?.name === stream.stream?.name && item.stream?.namespace === stream.stream?.namespace
+      )?.config?.selected !== stream.config?.selected;
+
+    const rowChanged = !isEqual(
+      initialValues.syncCatalog.streams.find(
+        (item) =>
+          item.stream &&
+          stream.stream &&
+          item.stream.name === stream.stream.name &&
+          item.stream.namespace === stream.stream.namespace
+      )?.config,
+      stream.config
+    );
+
+    if (rowStatusChanged) {
+      return isStreamEnabled ? "added" : "removed";
+    } else if (rowChanged) {
+      return "changed";
+    } else if (!isStreamEnabled && !rowStatusChanged) {
+      return "disabled";
+    }
+    return "unchanged";
+  }, [initialValues.syncCatalog.streams, isStreamEnabled, stream.config, stream.stream]);
+
+  const pillButtonVariant: PillButtonVariant = useMemo(() => {
+    if (statusToDisplay === "added" && !isSelected) {
+      return "green";
+    } else if (statusToDisplay === "removed" && !isSelected) {
+      return "red";
+    } else if (statusToDisplay === "changed" || isSelected) {
+      return "blue";
+    }
+    return "grey";
+  }, [isSelected, statusToDisplay]);
+
+  const statusIcon = useMemo(() => {
```

Comment:
It's possible to make `CatalogTreeTableRowIcon` its own component based on some items returned from this hook rather than having the hook return a ReactNode.  I didn't have strong feelings on it but am open to change this if it seems better to do so.  

I liked that this keeps all of the conditional pieces in one central spot.

Reply:
I think it might benefit from being a component than being part of the hook.